dev/system-test/firestore.ts

```diff
@@ -2321,3 +2324,271 @@ describe('Client initialization', () => {
     });
   }
 });
+
+describe('Bundle building', () => {
+  // Helper function for testing: converts "13{..}43{..}51{..}" to [[13,43,51],[obj,obj,obj]]
+  function bundleToLengthsAndObject(bundle: string): [number[], object[]] {
+    const isNum = (c: string) => !isNaN(Number(c));
+    let jsons = '';
+    let bracketDepth = 0;
+    let insertComma = false;
+
+    // Turn "13{..}43{..}51{..}" into "{..},{..},{..}"
+    for (const c of bundle) {
+      if (isNum(c) && bracketDepth === 0) {
+        if (insertComma) {
+          jsons += ',';
+          insertComma = false;
+        }
+      } else {
+        jsons += c;
+        if (c === '{') {
+          bracketDepth += 1;
+        } else if (c === '}') {
+          bracketDepth -= 1;
+          if (bracketDepth === 0) {
+            insertComma = true;
+          }
+        }
+      }
+    }
+
+    let lengths = '';
+    bracketDepth = 0;
+    // Turn "13{..}43{..}51{..}" into "13,43,51,"
+    for (const c of bundle) {
+      if (isNum(c) && bracketDepth === 0) {
+        lengths += c;
+      } else {
+        if (c === '{') {
+          bracketDepth += 1;
+        } else if (c === '}') {
+          bracketDepth -= 1;
+          if (bracketDepth === 0) {
+            lengths += ',';
+          }
+        }
+      }
+    }
+
+    return [
+      JSON.parse(`[${lengths.slice(0, lengths.length - 1)}]`),
+      JSON.parse(`[${jsons}]`),
+    ];
+  }
+
+  let firestore: Firestore;
+  let testCol: CollectionReference;
+
+  beforeEach(async () => {
+    firestore = new Firestore({});
+    testCol = getTestRoot(firestore);
+    const ref1 = testCol.doc('doc1');
+    const ref2 = testCol.doc('doc2');
+    const ref3 = testCol.doc('doc3');
+    const ref4 = testCol.doc('doc4');
+
+    await Promise.all([
+      ref1.set({name: '1', sort: 1, value: 'string value'}),
+      ref2.set({name: '2', sort: 2, value: 42}),
+      ref3.set({name: '3', sort: 3, value: {nested: 'nested value'}}),
+      ref4.set({name: '4', sort: 4, value: FieldValue.serverTimestamp()}),
+    ]);
+  });
+
+  afterEach(() => verifyInstance(firestore));
+
+  it('succeeds to read length prefixed json with testing function', () => {
+    const bundleString =
+      '47{"a": "string value"}53{"b": 123}454{"c": {"d": "nested value"}}';
+    const [lengths, objects] = bundleToLengthsAndObject(bundleString);
+    expect(lengths).to.deep.equal([47, 53, 454]);
+    expect(objects).to.deep.equal([
+      {a: 'string value'},
+      {b: 123},
+      {c: {d: 'nested value'}},
+    ]);
+  });
+
+  it('succeeds with document snapshots', async () => {
```

Comment:
All of these tests should be unit tests in their own file (bundle.ts). We can cherry-pick some tests and duplicate them as system tests.

We also need at least one test that consumes `stream()` directly.

Reply:
System tests are usually easier to write, but they:

- are harder to debug
- are not deterministic
- make it impossible to test some edge cases
- cannot be run without Production credentials
- don't count against code coverage

Please add unit tests :)

Reply:
Moved some to unit tests, thanks.